### PR TITLE
Add callback API to oe_log_message().

### PR DESCRIPTION
**Pull request by nonpolarity**
The implementation is based on the design https://github.com/openenclave/openenclave/blob/master/docs/DesignDocs/oe_log_message()_callback_proposal.md.
The sample is added to samples/log_callback.

Signed-off-by: Alvin Chen <alvin@chen.asia>

### CODE REVIEW

**Comment by nonpolarity** [comment]
bors try

File: host/traceh.c [hunk 1]
```diff
@@ -28,8 +29,9 @@
     sprintf(buffer, format, argument)
 #endif
 
-static char* _log_level_strings[OE_LOG_LEVEL_MAX] =
+const char* oe_log_level_strings[OE_LOG_LEVEL_MAX] =
```
Comment (dthaler) [inline host/traceh.c:32]:
```suggestion
const char* const oe_log_level_strings[OE_LOG_LEVEL_MAX] =
```

Explanation: https://www.geeksforgeeks.org/difference-const-char-p-char-const-p-const-char-const-p/ explains about const char*: "You cannot change the value pointed by ptr, but you can change the pointer itself."   Since the array itself should not change, use the 3rd approach there using this suggestion.

File: host/traceh.c [hunk 2]
```diff
@@ -235,26 +237,29 @@ static bool _escape_characters(
 static void _write_message_to_stream(
     FILE* stream,
     bool is_enclave,
-    const char* time,
+    struct tm* t,
```
Comment (dthaler) [inline host/traceh.c:240]:
```suggestion
    const struct tm* t,
```

File: host/traceh.c [hunk 3]
```diff
@@ -235,26 +237,29 @@ static bool _escape_characters(
...
+        oe_log_level_strings[level],
         (long long unsigned int)oe_thread_self(),
         message);
 }
 
 static void _write_custom_format_message_to_stream(
     FILE* stream,
     bool is_enclave,
-    const char* time,
+    struct tm* t,
```
Comment (dthaler) [inline host/traceh.c:262]:
```suggestion
    const struct tm* t,
```

File: include/openenclave/log.h [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    OE_LOG_LEVEL_NONE = 0,
+    OE_LOG_LEVEL_FATAL,
+    OE_LOG_LEVEL_ERROR,
+    OE_LOG_LEVEL_WARNING,
+    OE_LOG_LEVEL_INFO,
+    OE_LOG_LEVEL_VERBOSE,
+    OE_LOG_LEVEL_MAX
+} oe_log_level_t;
+
+extern const char* oe_log_level_strings[OE_LOG_LEVEL_MAX];
```
Comment (dthaler) [inline include/openenclave/log.h:20]:
```suggestion
extern const char* const oe_log_level_strings[OE_LOG_LEVEL_MAX];
```

File: include/openenclave/trace.h [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+#include <openenclave/log.h>
+#include <time.h>
+
+OE_EXTERNC_BEGIN
+
+typedef void (*oe_log_callback_t)(
+    void* context,
+    bool is_enclave,
+    struct tm* t,
```
Comment (dthaler) [inline include/openenclave/trace.h:15]:
```suggestion
    const struct tm* t,
```

File: samples/log_callback/README.md [hunk 6]
```diff
@@ -0,0 +1,147 @@
+# oe_log_message Callback Sample
+
+This sample demonstrates how to customize log function through the new callback feature of oe_log_message().
```
Comment (dthaler) [inline samples/log_callback/README.md:3]:
```suggestion
This sample demonstrates how to customize the log function through the new callback feature of oe_log_message().
```

File: samples/log_callback/README.md [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+There are several problems with the current logging mechanism that prevent end users from making full use of logging:
+
+- Message can't be redirected to different destinations on demand.
```
Comment (dthaler) [inline samples/log_callback/README.md:6]:
```suggestion
- Messages can't be redirected to different destinations on demand.
```

File: samples/log_callback/README.md [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+- Lacking of a method for end users to customize messages.
```
Comment (dthaler) [inline samples/log_callback/README.md:7]:
```suggestion
- There is no method for application developers to customize messages.
```

File: samples/log_callback/README.md [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+- It doesn't have a mechanism to communicate between the host application and oe_log_message() such that end users can't pass more information to the message function.
```
Comment (dthaler) [inline samples/log_callback/README.md:8]:
```suggestion
- It doesn't have a mechanism to communicate between the host application and oe_log_message() such that application developers can't pass more information to the message function.
```

File: samples/log_callback/README.md [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+
+The callback API is supposed to solve such problems.
+
+## User Experience
+
+By making use of the callback API of oe_log_message(), a user can customize the output of the log message:
```
Comment (dthaler) [inline samples/log_callback/README.md:14]:
```suggestion
By making use of the callback API of oe_log_message(), a developer can customize the output of the log message:
```

File: samples/log_callback/README.md [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+This sample demonstrates how to customize log function through the new callback feature of oe_log_message().
+There are several problems with the current logging mechanism that prevent end users from making full use of logging:
+
+- Message can't be redirected to different destinations on demand.
+- Lacking of a method for end users to customize messages.
+- It doesn't have a mechanism to communicate between the host application and oe_log_message() such that end users can't pass more information to the message function.
+
+The callback API is supposed to solve such problems.
+
+## User Experience
```
Comment (dthaler) [inline samples/log_callback/README.md:12]:
```suggestion
## Developer Experience
```

File: samples/log_callback/README.md [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+
+By making use of the callback API of oe_log_message(), a user can customize the output of the log message:
+
+- The output destination can be stdout, stderr, or any specified files that the destination is adjustable during runtime.
+- The message can be adjusted according to the log level.
+- The user can get at least as much information about the message as the internal API. For example, log time, enclave or host side, and log level.
+Finally, the user can define a function to customize the log message and can designate the funtion to process all the log messages.
```
Comment (dthaler) [inline samples/log_callback/README.md:19]:
```suggestion
- The developer can get at least as much information about the message as the internal API. For example, log time, enclave or host side, and log level.
Finally, the developer can define a function to customize the log message and can designate the function to process all the log messages.
```

File: samples/log_callback/README.md [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+
+## Changes
```
Comment (dthaler) [inline samples/log_callback/README.md:21]: I don't think this section is needed in a samples readme.  Only in the design doc.  Recommend removing.

File: samples/log_callback/README.md [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+    oe_log_level_t level,
+    uint64_t host_thread_id,
+    const char* message);
+oe_result_t oe_log_set_callback(void* context, oe_log_callback_t callback);
+```
+Since log.h is included by trace.h, the end user can perform customized logging through including trace.h.
+
+
+## How to customize the log
+To customize the log, the user needs to prepare a customized log processing function "customized_log()":
```
Comment (dthaler) [inline samples/log_callback/README.md:55]:
```suggestion
To customize the log, the developer needs to prepare a customized log processing function "customized_log()":
```

File: samples/log_callback/README.md [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+```
+void customized_log(
+    void* context,
+    bool is_enclave,
+    struct tm* t,
```
Comment (dthaler) [inline samples/log_callback/README.md:60]:
```suggestion
    const struct tm* t,
```

File: samples/log_callback/README.md [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+        time,
+        usecs,
+        (is_enclave ? "E" : "H"),
+        oe_log_level_strings[level],
+        host_thread_id,
+        message);
+}
+```
+which must follow the function type "*oe_log_callback_t". All these parameters are some information about logging,
+but the user has the full control to make use of them. The timestamp information, which is not printable,
```
Comment (dthaler) [inline samples/log_callback/README.md:91]:
```suggestion
but the developer has the full control to make use of them. The timestamp information, which is not printable,
```

File: samples/log_callback/README.md [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+
+To pass in the specified log file "./oe_out.txt", the file must be opened before registration.
+In this case the context is used as the logging destination. Besides that, context can be casted back as any desired type, since type of context is "void*".
+```
+    FILE* out_file = fopen("./oe_out.txt", "w");
+    oe_log_set_callback((void*)out_file, customized_log);
+
+```
+When oe_log_message() is called, the most recently registered log function will be invoked. Obviously, if it is registered before starting the enclave, all the logs are affected.
+In the user's expectation, log messages with level equal or greater than OE_LOG_LEVEL_WARNING will be writen to "./oe_out.txt", and others will be writen to "stderr". The typical output is as following:
```
Comment (dthaler) [inline samples/log_callback/README.md:105]:
```suggestion
In the developer's expectation, log messages with level equal or greater than OE_LOG_LEVEL_WARNING will be written to "./oe_out.txt", and others will be written to "stderr". The typical output is as following:
```

File: samples/log_callback/README.md [hunk 18]
```diff
@@ -0,0 +1,147 @@
...
+
+```bash
+cd log_callback
+make build
+make run
+```
+
+#### Note
+
+log_callback sample can run under Open Enclave simulation mode.
```
Comment (dthaler) [inline samples/log_callback/README.md:138]:
```suggestion
The log_callback sample can run under Open Enclave simulation mode.
```

File: samples/log_callback/README.md [hunk 19]
```diff
@@ -0,0 +1,147 @@
...
+
+To run the switchless sample in simulation mode from the command like, use the following:
```
Comment (dthaler) [inline samples/log_callback/README.md:140]: This isn't the "switchless" sample

File: samples/log_callback/host/host.c [hunk 20]
```diff
@@ -0,0 +1,126 @@
...
+// print a message.
+void host_log_callback()
+{
+    fprintf(stdout, "Enclave called into host to print: Hello World!\n");
+}
+
+void customized_log(
+    void* context,
+    bool is_enclave,
+    struct tm* t,
```
Comment (dthaler) [inline samples/log_callback/host/host.c:38]:
```suggestion
    const struct tm* t,
```

File: samples/test-samples.cmake [hunk 21]
```diff
@@ -40,14 +40,12 @@ endif ()
 
 # Set SAMPLES_LIST so that helloworld becomes the first if BUILD_ENCLAVES=ON.
 if (BUILD_ENCLAVES)
-  set(SAMPLES_LIST helloworld file-encryptor switchless host_verify)
```
Comment (dthaler) [inline samples/test-samples.cmake:43]: Why was host_verify removed here?

File: samples/log_callback/README.md (hunk 19 above)
Comment (nonpolarity) [reply samples/log_callback/README.md]: Fixed. Thanks for all the suggestions and pointing out the error!

File: samples/test-samples.cmake (hunk 21 above)
Comment (nonpolarity) [reply samples/test-samples.cmake:43]: Sorry to make you confused. There is a PR https://github.com/openenclave/openenclave/pull/3463 which may remove it before merging.

Comment (dthaler) [reply samples/test-samples.cmake:43]: https://github.com/openenclave/openenclave/blob/master/docs/Community/Contributing.md says:

> **DON'T** mix independent, unrelated changes in one PR.

File: samples/log_callback/README.md (hunk 13 above)
Comment (nonpolarity) [reply samples/log_callback/README.md]: Removed, Thanks!

File: samples/test-samples.cmake (hunk 21 above)
Comment (nonpolarity) [reply samples/test-samples.cmake:43]: Removed, thanks!

File: samples/test-samples.cmake [hunk 22]
```diff
@@ -40,7 +40,8 @@ endif ()
...
+  set(SAMPLES_LIST file-encryptor helloworld host_verify log_callback
```
Comment (radhikaj) [inline samples/test-samples.cmake:43]: Can you add log_callback to the samples readme?

Comment (radhikaj) [reply samples/test-samples.cmake:43]: https://github.com/openenclave/openenclave/blob/master/samples/README_Linux.md and README_Windows.md

File: samples/test-samples.cmake [hunk 23]
```diff
@@ -50,6 +51,15 @@ else ()
...
+if (BUILD_ENCLAVES)
+  set(SAMPLES_LIST helloworld file-encryptor log_callback switchless)
+  # Debug malloc will set allocated memory to a fixed pattern.
+  # Hence do not enable pluggable_allocator test under USE_DEBUG_MALLOC.
+  if (COMPILER_SUPPORTS_SNMALLOC AND NOT USE_DEBUG_MALLOC)
+    list(APPEND SAMPLES_LIST pluggable_allocator)
+  endif ()
+endif ()
+
 if ($ENV{OE_SIMULATION})
```
Comment (radhikaj) [inline samples/test-samples.cmake:63]: Can you verify that the sample gets packed and installed? Thanks

Comment (radhikaj) [reply samples/test-samples.cmake:63]: Also add this sample to https://github.com/openenclave/openenclave/blob/ac4169d199906905414980a7fad60a56f16fe58a/docs/GettingStartedDocs/install_oe_sdk-Simulation.md#considerations

File: samples/log_callback/README.md [hunk 24]
```diff
@@ -0,0 +1,114 @@
...
+
+The callback API is supposed to solve such problems.
+
+## Developer Experience
+
+By making use of the callback API of oe_log_message(), a developer can customize the output of the log message:
+
+- The output destination can be stdout, stderr, or any specified files that the destination is adjustable during runtime.
+- The message can be adjusted according to the log level.
+- The developer can get at least as much information about the message as the internal API. For example, log time, enclave or host side, and log level.
```
Comment (radhikaj) [inline samples/log_callback/README.md:18]:
```suggestion
- The developer gets the following information- log time, enclave or host side, and log level.
```

File: samples/log_callback/README.md [hunk 25]
```diff
@@ -0,0 +1,114 @@
...
+- There is no method for application developers to customize messages.
+- It doesn't have a mechanism to communicate between the host application and oe_log_message() such that application developers can't pass more information to the message function.
+
+The callback API is supposed to solve such problems.
+
+## Developer Experience
+
+By making use of the callback API of oe_log_message(), a developer can customize the output of the log message:
+
+- The output destination can be stdout, stderr, or any specified files that the destination is adjustable during runtime.
```
Comment (radhikaj) [inline samples/log_callback/README.md:16]:
```suggestion
- The output destination can be stdout, stderr or any file. The destination is adjustable during runtime.
```

File: samples/log_callback/README.md [hunk 26]
```diff
@@ -0,0 +1,114 @@
...
+In this case the context is used as the logging destination. Besides that, context can be casted back as any desired type, since type of context is "void*".
+```
+    FILE* out_file = fopen("./oe_out.txt", "w");
+    oe_log_set_callback((void*)out_file, customized_log);
+
+```
+When oe_log_message() is called, the most recently registered log function will be invoked. Obviously, if it is registered before starting the enclave, all the logs are affected.
+In the developer's expectation, log messages with level equal or greater than OE_LOG_LEVEL_WARNING will be written to "./oe_out.txt", and others will be written to "stderr". The typical output is as following:
+```
+alvin@wechen3-u18-2:~/openenclave/build/samples/log_callback$ cat oe_out.txt
```
Comment (radhikaj) [inline samples/log_callback/README.md:74]: Maybe remove your name and id, just having $cat oe_out.txt is enough.

File: include/openenclave/internal/trace.h [hunk 27]
```diff
@@ -7,39 +7,18 @@
 #include <openenclave/bits/defs.h>
 #include <openenclave/bits/result.h>
 #include <openenclave/bits/types.h>
+//#include <openenclave/trace.h>
```
Comment (achamayou) [inline include/openenclave/internal/trace.h:10]:
```suggestion
```

**Review by achamayou — APPROVED** [review]

File: include/openenclave/log.h [hunk 28]
```diff
@@ -0,0 +1,24 @@
+// Copyright (c) Open Enclave SDK contributors.
+// Licensed under the MIT License.
+
+#ifndef _OE_LOG_USER
```
Comment (mingweishih) [inline include/openenclave/log.h:4]:
```suggestion
#ifndef _OE_LOG_H
```

File: include/openenclave/log.h [hunk 29]
```diff
@@ -0,0 +1,24 @@
...
+#define _OE_LOG_USER
```
Comment (mingweishih) [inline include/openenclave/log.h:5]:
```suggestion
#define _OE_LOG_H
```

File: include/openenclave/trace.h [hunk 30]
```diff
@@ -0,0 +1,24 @@
+// Copyright (c) Open Enclave SDK contributors.
+// Licensed under the MIT License.
+
+#ifndef _OE_TRACE_USER
+#define _OE_TRACE_USER
```
Comment (mingweishih) [inline include/openenclave/trace.h:5]:
```suggestion
#define _OE_TRACE_H
```

File: samples/log_callback/CMakeLists.txt [hunk 31]
```diff
@@ -0,0 +1,48 @@
...
+
+cmake_minimum_required(VERSION 3.11)
+
+if (LVI_MITIGATION MATCHES ControlFlow)
+  # Configure the cmake to use customized compilation toolchain.
+  # This package has to be added before `project()`.
+  find_package(OpenEnclave-LVI-Mitigation CONFIG REQUIRED)
+endif ()
+
+project("Hello World Sample" LANGUAGES C CXX)
```
Comment (mingweishih) [inline samples/log_callback/CMakeLists.txt:12]: Update the project name

File: samples/log_callback/README.md [hunk 32]
```diff
@@ -0,0 +1,114 @@
+# oe_log_message Callback Sample
+
+This sample demonstrates how to customize the log function through the new callback feature of oe_log_message().
```
Comment (mingweishih) [inline samples/log_callback/README.md:3]: I would just discard new.

```suggestion
This sample demonstrates how to customize the log function through the callback feature of oe_log_message().
```

File: samples/log_callback/README.md [hunk 33]
```diff
@@ -0,0 +1,114 @@
...
+There are several problems with the current logging mechanism that prevent end users from making full use of logging:
```
Comment (mingweishih) [inline samples/log_callback/README.md:4]: I don't think we need to explain things here, which are already part of the design doc. I'd suggest remove line 4-10, just directly start with Developer Experience that explains how the callback work.

Alternatively, you could specify the version of OE. For example,

This sample demonstrates how to customize the log function through the new callback feature of oe_log_message() introduced in v0.12.

There are several problems with the current logging mechanism (prior to v0.12) that prevent end users from making full use of logging:
...

File: samples/log_callback/README.md (hunk 26 above)
Comment (nonpolarity) [reply samples/log_callback/README.md]: Removed, thanks!

File: samples/log_callback/CMakeLists.txt (hunk 31 above)
Comment (nonpolarity) [reply samples/log_callback/CMakeLists.txt]: Updated, thanks!

File: samples/log_callback/README.md (hunk 33 above)
Comment (nonpolarity) [reply samples/log_callback/README.md:4]: Thanks for the suggestion, yes it is not good to put it in sample. Removed.

File: include/openenclave/trace.h [hunk 34]
```diff
@@ -0,0 +1,24 @@
+// Copyright (c) Open Enclave SDK contributors.
+// Licensed under the MIT License.
+
+#ifndef _OE_TRACE_USER
```
[the remark this answers is not in the document]
Comment (nonpolarity) [reply include/openenclave/trace.h]: Thanks for the suggestion, to avoid conflict with include/openenclave/internal/traceh.h I'd like to rename it as _OE_TRACE_USER_H

[the remark this answers is not in the document]
Comment (dthaler) [reply include/openenclave/trace.h]: My opinion: It should match the filename, so this should be _OE_TRACE_H
internal/traceh.h should be _OE_INTERNAL_TRACEH_H

File: samples/test-samples.cmake [hunk 35]
```diff
@@ -40,7 +40,16 @@ endif ()
 
 # Set SAMPLES_LIST so that helloworld becomes the first if BUILD_ENCLAVES=ON.
 if (BUILD_ENCLAVES)
-  set(SAMPLES_LIST helloworld file-encryptor switchless)
+  set(SAMPLES_LIST file-encryptor helloworld log_callback switchless)
+  # Debug malloc will set allocated memory to a fixed pattern.
+  # Hence do not enable pluggable_allocator test under USE_DEBUG_MALLOC.
+  if (COMPILER_SUPPORTS_SNMALLOC AND NOT USE_DEBUG_MALLOC)
+    list(APPEND SAMPLES_LIST pluggable_allocator)
+  endif ()
```
Comment (dthaler) [inline samples/test-samples.cmake:48]: Why are lines 44-48 needed in this PR?  They look to be orthogonal to the log_message support.

File: include/openenclave/trace.h (hunk 34 above)
[the remark this answers is not in the document]
Comment (nonpolarity) [reply include/openenclave/trace.h]: Sorry I have a typo there. Macro in include/openenclave/internal/trace.h is changed to _OE_INTERNAL_TRACE_H

File: samples/test-samples.cmake (hunk 35 above)
Comment (nonpolarity) [reply samples/test-samples.cmake]: Removed, thanks!

**Comment by nonpolarity** [comment]
bors try

**Review by dthaler — APPROVED** [review]

File: samples/test-samples.cmake (hunk 22 above)
Comment (nonpolarity) [reply samples/test-samples.cmake]: Thanks, now it is updated.

File: samples/log_callback/log_callback.edl [hunk 36]
```diff
@@ -0,0 +1,15 @@
+// Copyright (c) Open Enclave SDK contributors.
+// Licensed under the MIT License.
+
+enclave {
+    from "openenclave/edl/syscall.edl" import *;
+    from "platform.edl" import *;
+
+    trusted {
+        public void enclave_log_callback();
```
Comment (anakrish) [inline samples/log_callback/log_callback.edl:9]: This ecall and the ocall below are themselves not logging callbacks. Therefore the names are misleading.
You could just call them: enc_hello and host_hello.

**Review by anakrish — APPROVED** [review]
LGTM.
It is better to rename the edl ecall and ocall since they are not logging callbacks.

File: samples/log_callback/log_callback.edl (hunk 36 above)
Comment (nonpolarity) [reply samples/log_callback/log_callback.edl]: Thanks for the suggestion, both are fixed.

**Comment by nonpolarity** [comment]
bors try

**Review by anakrish — APPROVED** [review]

**Comment by nonpolarity** [comment]
bors try
